File: src/EditorFeatures/CSharpTest/Outlining/MetadataAsSource/MethodDeclarationOutlinerTests.cs
// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System.Collections.Generic;
using System.Threading;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Editor.CSharp.Outlining;
using Microsoft.CodeAnalysis.Editor.Implementation.Outlining;
using Microsoft.CodeAnalysis.Text;
using Roslyn.Test.Utilities;
using Roslyn.Utilities;
using Xunit;
using MaSOutliners = Microsoft.CodeAnalysis.Editor.CSharp.Outlining.MetadataAsSource;

namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Outlining.MetadataAsSource
{
    public class MethodDeclarationOutlinerTests :
        AbstractOutlinerTests<MethodDeclarationSyntax>
    {
        internal override IEnumerable<OutliningSpan> GetRegions(MethodDeclarationSyntax node)
        {
            var outliner = new MaSOutliners.MethodDeclarationOutliner();
            return outliner.GetOutliningSpans(node, CancellationToken.None).WhereNotNull();
        }

        [WpfFact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)]
        public void NoCommentsOrAttributes()
        {
            var tree = ParseCode(
@"class Foo
{
    public string Bar(int x);
}");
            var typeDecl = tree.DigToFirstTypeDeclaration();
            var method = typeDecl.DigToFirstNodeOfType<MethodDeclarationSyntax>();

            Assert.Empty(GetRegions(method));
        }

        [WpfFact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)]
        public void WithAttributes()
        {
            var tree = ParseCode(
@"class Foo
{
    [Foo]
    public string Bar(int x);
}");
            var typeDecl = tree.DigToFirstTypeDeclaration();
            var method = typeDecl.DigToFirstNodeOfType<MethodDeclarationSyntax>();

            var actualRegion = GetRegion(method);
            var expectedRegion = new OutliningSpan(
                TextSpan.FromBounds(18, 29),
                TextSpan.FromBounds(18, 54),
                CSharpOutliningHelpers.Ellipsis,
                autoCollapse: true);

            AssertRegion(expectedRegion, actualRegion);
        }

        [WpfFact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)]
        public void WithCommentsAndAttributes()
        {
            var tree = ParseCode(
@"class Foo
{
    // Summary:
    //     This is a summary.
    [Foo]
    string Bar(int x);
}");
            var typeDecl = tree.DigToFirstTypeDeclaration();
            var method = typeDecl.DigToFirstNodeOfType<MethodDeclarationSyntax>();

            var actualRegion = GetRegion(method);
            var expectedRegion = new OutliningSpan(
                TextSpan.FromBounds(18, 77),
                TextSpan.FromBounds(18, 95),
                CSharpOutliningHelpers.Ellipsis,
                autoCollapse: true);

            AssertRegion(expectedRegion, actualRegion);
        }

        [WpfFact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)]
        public void WithCommentsAttributesAndModifiers()
        {
            var tree = ParseCode(
@"class Foo
{
    // Summary:
    //     This is a summary.
    [Foo]
    public string Bar(int x);
}");
            var typeDecl = tree.DigToFirstTypeDeclaration();
            var method = typeDecl.DigToFirstNodeOfType<MethodDeclarationSyntax>();

            var actualRegion = GetRegion(method);
            var expectedRegion = new OutliningSpan(
                TextSpan.FromBounds(18, 77),
                TextSpan.FromBounds(18, 102),
                CSharpOutliningHelpers.Ellipsis,
                autoCollapse: true);

            AssertRegion(expectedRegion, actualRegion);
        }
    }
}

File: src/EditorFeatures/VisualBasicTest/Outlining/MetadataAsSource/MethodDeclarationOutlinerTests.vb
' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports System.Threading
Imports Microsoft.CodeAnalysis.Editor.VisualBasic.Outlining
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports MaSOutliners = Microsoft.CodeAnalysis.Editor.VisualBasic.Outlining.MetadataAsSource
Imports Microsoft.CodeAnalysis.Editor.Implementation.Outlining

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.Outlining.MetadataAsSource
    Public Class MethodDeclarationOutlinerTests
        Inherits AbstractVisualBasicSyntaxOutlinerTests(Of MethodStatementSyntax)

        Friend Overrides Function GetRegions(node As MethodStatementSyntax) As IEnumerable(Of OutliningSpan)
            Dim outliner = New MaSOutliners.MethodDeclarationOutliner()
            Return outliner.GetOutliningSpans(node, CancellationToken.None).WhereNotNull()
        End Function

        Private Shared Function GetMethodStatement(code As Xml.Linq.XElement) As MethodStatementSyntax
            Dim tree = ParseCode(code.Value)
            Dim typeDecl = tree.DigToFirstTypeBlock()
            Dim methodBlock = typeDecl.DigToFirstNodeOfType(Of MethodBlockSyntax)()
            Dim methodStatement = DirectCast(methodBlock.BlockStatement, MethodStatementSyntax)
            Return methodStatement
        End Function

        <WpfFact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)>
        Public Sub NoCommentsOrAttributes()
            Dim code =
<code><![CDATA[
Class C
    Sub M()
    End Sub
End Class
]]></code>

            Dim methodStatement As MethodStatementSyntax = GetMethodStatement(code)

            Assert.Empty(GetRegions(methodStatement))
        End Sub

        <WpfFact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)>
        Public Sub WithAttributes()
            Dim code =
<code><![CDATA[
Class C
    <Foo>
    Sub M()
    End Sub
End Class
]]></code>

            Dim methodStatement As MethodStatementSyntax = GetMethodStatement(code)

            Dim actualRegion = GetRegion(methodStatement)
            Dim expectedRegion = New OutliningSpan(
                TextSpan.FromBounds(13, 23),
                TextSpan.FromBounds(13, 30),
                VisualBasicOutliningHelpers.Ellipsis,
                autoCollapse:=True)

            AssertRegion(expectedRegion, actualRegion)
        End Sub

        <WpfFact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)>
        Public Sub WithCommentsAndAttributes()
            Dim code =
<code><![CDATA[
Class C
    ' Summary:
    '     This is a summary.
    Sub M()
    End Sub
End Class
]]></code>

            Dim methodStatement As MethodStatementSyntax = GetMethodStatement(code)

            Dim actualRegion = GetRegion(methodStatement)
            Dim expectedRegion = New OutliningSpan(
                TextSpan.FromBounds(13, 57),
                TextSpan.FromBounds(13, 64),
                VisualBasicOutliningHelpers.Ellipsis,
                autoCollapse:=True)

            AssertRegion(expectedRegion, actualRegion)
        End Sub

        <WpfFact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)>
        Public Sub WithCommentsAttributesAndModifiers()
            Dim code =
<code><![CDATA[
Class C
    ' Summary:
    '     This is a summary.
    Public Sub M()
    End Sub
End Class
]]></code>

            Dim methodStatement As MethodStatementSyntax = GetMethodStatement(code)

            Dim actualRegion = GetRegion(methodStatement)
            Dim expectedRegion = New OutliningSpan(
                TextSpan.FromBounds(13, 57),
                TextSpan.FromBounds(13, 71),
                VisualBasicOutliningHelpers.Ellipsis,
                autoCollapse:=True)

            AssertRegion(expectedRegion, actualRegion)
        End Sub
    End Class
End Namespace
